src/backend/distributed/executor/multi_utility.c

```diff
@@ -816,6 +832,131 @@ ProcessAlterObjectSchemaStmt(AlterObjectSchemaStmt *alterObjectSchemaStmt,
 }
 
 
+/* TODO: Write function comments */
+static Node *
+ProcessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand, bool isTopLevel)
+{
+	Oid relationId = InvalidOid;
+	List *taskList = NIL;
+
+	if (vacuumStmt->relation == NULL)
+	{
+		return (Node *) vacuumStmt;
+	}
+
+	relationId = RangeVarGetRelid(vacuumStmt->relation, NoLock, false);
+
+	/* first check whether a distributed relation is affected */
+	if (!OidIsValid(relationId) || !IsDistributedTable(relationId))
+	{
+		return (Node *) vacuumStmt;
+	}
+
+	taskList = VacuumTaskList(relationId, vacuumStmt);
+
+	SavedMultiShardCommitProtocol = MultiShardCommitProtocol;
+	MultiShardCommitProtocol = COMMIT_PROTOCOL_BARE;
+	ExecuteModifyTasksWithoutResults(taskList);
+
+	return (Node *) vacuumStmt;
+}
+
+
+/* TODO: Write function comments */
+static List *
+VacuumTaskList(Oid relationId, VacuumStmt *vacuumStmt)
+{
+	List *taskList = NIL;
+	List *shardIntervalList = LoadShardIntervalList(relationId);
+	ListCell *shardIntervalCell = NULL;
+	uint64 jobId = INVALID_JOB_ID;
+	int taskId = 1;
+	StringInfo vacuumString = DeparseVacuumStmtPrefix(vacuumStmt);
+	const int vacuumPrefixLen = vacuumString->len;
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	char *tableName = get_rel_name(relationId);
+
+	/* lock metadata before getting placement lists */
+	LockShardListMetadata(shardIntervalList, ExclusiveLock);
```

Comment:
Without locks - as it stands right now - the relation could have vanished in any of the preceding lines. That's not good.